internal/lib/conmonmon.go

```diff
@@ -0,0 +1,183 @@
+package lib
+
+import (
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/containers/psgo"
+	"github.com/cri-o/cri-o/internal/oci"
+	"github.com/cri-o/cri-o/pkg/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+// the time between checking the registered conmons
+var sleepTime = 5 * time.Minute
+
+// conmonmon is a struct responsible for monitoring conmons
+// it contains a map of containers -> info on conmons, and sleeps on
+// a loop, waiting for a conmon to die. if it has, it kills the associated
+// container.
+type conmonmon struct {
+	conmons   map[*oci.Container]*conmonInfo
+	closeChan chan bool
+	runtime   *oci.Runtime
+	server    *ContainerServer
+	lock      sync.Mutex
+}
+
+// conmonInfo contains all necessary state to verify
+// the conmon the container was originally spawned from is still running
+type conmonInfo struct {
+	conmonPID int
+	startTime string
+}
+
+// newConmonmon creates a new conmonmon instance given a runtime.
+// It also starts the monitoring routine
+func (c *ContainerServer) newConmonmon(r *oci.Runtime) *conmonmon {
+	cmm := conmonmon{
+		conmons:   make(map[*oci.Container]*conmonInfo),
+		runtime:   r,
+		server:    c,
+		closeChan: make(chan bool, 2),
+	}
+	go cmm.monitorConmons()
+	return &cmm
+}
+
+// monitorConmons sits on a loop and sleeps.
+// after waking, it signals to the conmons, checking if they're still alive.
+func (c *conmonmon) monitorConmons() {
+	for {
+		select {
+		case <-time.After(sleepTime):
+			c.signalConmons()
+		case <-c.closeChan:
+			return
+		}
+	}
+}
+
+// signalConmons loops through all available conmons and they are verified as still alive
+// if they're not, the container is killed and we spoof an OOM event for the container
+func (c *conmonmon) signalConmons() {
+	c.lock.Lock()
+	for ctr, info := range c.conmons {
+		if ctr.State().Status == oci.ContainerStateRunning {
+			if err := c.verifyConmonValid(info.startTime, info.conmonPID); err != nil {
+				logrus.Debugf("conmon pid %d invalid: %v. Killing container %s", info.conmonPID, err, ctr.ID())
+				delete(c.conmons, ctr)
+				// kill container in separate thread to hold the conmonmon lock as little as possible
+				go c.oomKillContainer(ctr)
+			}
+		}
+	}
+	c.lock.Unlock()
+}
+
+// verifyConmonValid checks if the conmon we have saved as being associated with the container
+// matches the pid. We first check if the pid is running, then check against the start time
+// we originally recorded.
+// These two checks should verify we are looking at the same conmon
+func (c *conmonmon) verifyConmonValid(savedStart string, pid int) error {
+	// verify it's still a running pid
+	if err := unix.Kill(pid, 0); err != nil {
```

Comment:
I think we can avoid the Kill completely, if the pid doesn't exist we will catch the error later

Reply:
Yeah agree :+1:

Reply:
On the other hand, wouldn't kill be cheaper if we don't find a matching pid number to avoid having to read and compare the start time?

Reply:
Not in the standard case. For most of the queries (the original conmon is still running), we check both (to make sure the pid that we sent 0 to is the one we originally thought). Thus, removing the kill takes an operation away in the common case, and adds an unnecessary disk hit in the special case (conmon was killed). I think we are right to leave it out